evap/evaluation/models_logging.py

```diff
@@ -0,0 +1,290 @@
+            items = [str(obj) for obj in related_objects] + ["�"] * missing
+            yield FieldAction(label, field_action_type, items)
+        elif hasattr(field, "choices") and field.choices:
+            # convert values from choice-based fields to their display equivalent
+            yield FieldAction(label, field_action_type, [_choice_to_display(field, item) for item in items])
+        elif isinstance(field, models.BooleanField):
+            # convert boolean to yes/no
+            yield FieldAction(label, field_action_type, list(map(yesno, items)))
+        else:
+            # as a default, just use plain items
+            yield FieldAction(label, field_action_type, items)
+
+
+class LogEntry(models.Model):
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="logs_about_me")
+    content_object_id = models.PositiveIntegerField(db_index=True)
+    content_object = GenericForeignKey("content_type", "content_object_id")
+    attached_to_object_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="logs_for_me")
+    attached_to_object_id = models.PositiveIntegerField(db_index=True)
+    attached_to_object = GenericForeignKey("attached_to_object_type", "attached_to_object_id")
+    datetime = models.DateTimeField(auto_now_add=True)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, null=True, blank=True, on_delete=models.PROTECT)
+    action_type = models.CharField(max_length=255, choices=zip(*itertools.tee(InstanceActionType)))
+    request_id = models.CharField(max_length=36, null=True, blank=True)
+    data = JSONField(default=dict, encoder=LogJSONEncoder)
+
+    class Meta:
+        ordering = ("-datetime", "-id")
+
+    def _evaluation_log_template_context(self, data):
+        model = self.content_type.model_class()
+        return {
+            field_name: list(_field_actions_for_field(model, field_name, actions))
+            for field_name, actions in data.items()
+        }
+
+    def display(self):
+        if self.action_type == InstanceActionType.CHANGE:
+            if self.content_object:
+                message = _("The {cls} {obj} was changed.")
+            else:
+                message = _("A {cls} was changed.")
+        elif self.action_type == InstanceActionType.CREATE:
+            if self.content_object:
+                message = _("The {cls} {obj} was created.")
+            else:
+                message = _("A {cls} was created.")
+        elif self.action_type == InstanceActionType.DELETE:
+            message = _("A {cls} was deleted.")
+
+        message = message.format(
+            cls=self.content_type.model_class()._meta.verbose_name_raw,
+            obj=f"\"{self.content_object!s}\"" if self.content_object else "",
+        )
+
+        return render_to_string("log/changed_fields_entry.html", {
+            'message': message,
+            'fields': self._evaluation_log_template_context(self.data),
```

Comment:
a method that's only called from a template and returns a rendered template smells to me like it should be template include. `message` and `fields` could be properties on `LogEntry` (perhaps with a more descriptive name for `fields`)

Reply:
You are right indeed. I changed that.